packages/react-reconciler/src/SchedulingProfiler.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {Lane, Lanes} from './ReactFiberLane';
+import type {Fiber} from './ReactInternalTypes';
+import type {Wakeable} from 'shared/ReactTypes';
+
+import {enableSchedulingProfiler} from 'shared/ReactFeatureFlags';
+import getComponentName from 'shared/getComponentName';
+import {getStackByFiberInDevAndProd} from './ReactFiberComponentStack';
+
+/**
+ * If performance exists and supports the subset of the User Timing API that we
+ * require.
+ */
+const supportsUserTiming =
+  typeof performance !== 'undefined' && typeof performance.mark === 'function';
+
+function formatLanes(laneOrLanes: Lane | Lanes): string {
+  return ((laneOrLanes: any): number).toString();
+}
+
+export function markCommitStarted(lanes: Lanes): void {
+  if (enableSchedulingProfiler) {
+    if (supportsUserTiming) {
+      performance.mark(`--commit-start-${formatLanes(lanes)}`);
+    }
+  }
+}
+
+export function markCommitStopped(): void {
+  if (enableSchedulingProfiler) {
+    if (supportsUserTiming) {
+      performance.mark(`--commit-stop`);
+    }
+  }
+}
+
+const PossiblyWeakMap = typeof WeakMap === 'function' ? WeakMap : Map;
+
+// $FlowFixMe: Flow cannot handle polymorphic WeakMaps
+const wakeableIDs: WeakMap<Wakeable, number> = new PossiblyWeakMap();
+let wakeableID: number = 0;
+function getWakeableID(wakeable: Wakeable): number {
+  if (!wakeableIDs.has(wakeable)) {
+    wakeableIDs.set(wakeable, wakeableID++);
+  }
+  return ((wakeableIDs.get(wakeable): any): number);
+}
+
+// $FlowFixMe: Flow cannot handle polymorphic WeakMaps
+const cachedFiberStacks: WeakMap<Fiber, string> = new PossiblyWeakMap();
+function cacheFirstGetComponentStackByFiber(workInProgress: Fiber): string {
+  if (!cachedFiberStacks.has(workInProgress)) {
+    const componentStack = getStackByFiberInDevAndProd(workInProgress) || '';
+    cachedFiberStacks.set(workInProgress, componentStack);
+  }
+  return ((cachedFiberStacks.get(workInProgress): any): string);
+}
```

Comment:
Worth checking to see if we cached the alternate too! (Double check my code below, didn't test it.)
```suggestion
function cacheFirstGetComponentStackByFiber(fiber: Fiber): string {
  if (cachedFiberStacks.has(fiber)) {
    return ((cachedFiberStacks.get(fiber): any): string);
  } else {
    const alternate = fiber.alternate;
    if (alternate !== null && cachedFiberStacks.has(alternate)) {
      return ((cachedFiberStacks.get(alternate): any): string);
    }
  }
  const componentStack = getStackByFiberInDevAndProd(fiber) || "";
  cachedFiberStacks.set(fiber, componentStack);
  return componentStack;
}
```